### PR TITLE
[Bugfix][Evacuation Controller]: Make sure that the number of VMs to evacuate is never negative

### PR DESCRIPTION
**Pull request by iholder101**
**What this PR does / why we need it**:
Lately we encountered a bug where the number of VMs to evacuate was negative. This was probably caused by some kind of a race.

For example, if there are 123 migrations running on a node, and ParallelOutboundMigrationsPerNode is then changed to 1, than the free spots of migrations to create in this node is: `ParallelOutboundMigrationsPerNode-123 = 1-123 < 0`.

This should never happen. Therefore, this variable is always set to 0 if negative.

**Which issue(s) this PR fixes**:
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=2212198

**Release note**:

```release-note
[Bugfix][Evacuation Controller]: Make sure that the number of VMs to evacuate is never negative
```

### CODE REVIEW

File: pkg/virt-controller/watch/drain/evacuation/evacuation.go [hunk 1]
```diff
@@ -403,8 +403,10 @@ func (c *EvacuationController) sync(node *k8sv1.Node, vmisOnNode []*virtv1.Virtu
 		return nil
 	}
 
-	diff := int(math.Min(float64(freeSpots), float64(len(migrationCandidates))))
-	remaining := freeSpots - diff
+	vmsToEvacuate := int(math.Min(float64(freeSpots), float64(len(migrationCandidates))))
```
Comment (xpivarc) [inline pkg/virt-controller/watch/drain/evacuation/evacuation.go:406]: How can this happen?
`if freeSpots <= 0 {
		c.Queue.AddAfter(node.Name, 5*time.Second)
		return nil
	}`
leaving only `migrationCandidates` to be the suspect. But that doesn't seem to be able to be `<0`.

I think the version of Kubevirt in the bug does not include https://github.com/kubevirt/kubevirt/commit/5d1b049a5154c72e1b888da5ca392a9b97858995

**Review by xpivarc — CHANGES_REQUESTED** [review]

File: pkg/virt-controller/watch/drain/evacuation/evacuation.go (hunk 1 above)
Comment (iholder101) [reply pkg/virt-controller/watch/drain/evacuation/evacuation.go:406]: Yup, you're right! Thanks!

Do you think it makes sense to keep this as another protection layer?
I mean, if we have another bug in this area, we at least won't panic.

WDYT?

Comment (xpivarc) [reply pkg/virt-controller/watch/drain/evacuation/evacuation.go:406]: I think for this case there are already unit tests present. I am definitely for improving coverage if anything. It is hard to argue about the need for this additional line as someone as well can argue it is redundant and therefore propose removal of it.

Comment (iholder101) [reply pkg/virt-controller/watch/drain/evacuation/evacuation.go:406]: Yeah, makes sense. Alright, if it's already taken care of with unit tests I think we can close this for now.
/close

**Pull request closed without merging**